src/chorus-one/physical-location.md

```diff
@@ -0,0 +1,515 @@
+   For redundancy,
+   we ensure that we have multiple machines that are _capable_ of being a validator at any time,
+   but at a given moment,
+   only a single instance is validating.
+
+While one could argue that validation happens at the validator machine,
+it is opaque to other peers in the network where our validator machine is located.
+Externally, only the sentry nodes are visible,
+and we can promote different machines to being validator without externally visible effects.
+Arguably,
+validation is the combination of a validator node and sentry nodes:
+without sentry nodes, the validator node is useless.
+
+To complicate matters further,
+while the validator node software produces blocks and votes,
+it does not sign them directly.
+At Chorus One we use hardware security modules (HSMs) for signing where possible.
+The hardware security modules we work with are external USB devices
+connected to a server.
+In the past we connected them directly to the machine
+where the validator node software runs,
+but this has a major downside:
+it makes it difficult to move workloads around.
+We have to ask data center personnel to unplug the HSM,
+and attach it to a different machine,
+which can take hours,
+or even weeks when shipping to a different country.
+
+To mitigate this, we use _remote signing_.
+We designate some machines as _signers_,
+and connect the HSMs to them.
+Validators then connect to the signers
+over a secure internal network to request a signature.
+For stability and performance reasons,
+the signer machines need to be reasonably close to validators,
+but they need not be in the same data center,
+and we have cases where they are not even in the same city.
+
+Because signing is essential for validation,
+we cannot allow signer machines to be a single point of failure,
+so we have multiple signers
+that are able to sign for a particular validator identity.
+When multiple signers are online,
+a signing request may be serviced by any of them.
+While a particular signature is always created by a particular signer,
+it might be that the vote for block _n_ and block _n_ + 1
+are signed by different signers.
+Therefore,
+even over a short time period,
+the question “where was the signer?”
+may not have a more precise answer than “in all of these cities”.
+
+To summarize,
+our setup for validating Cosmos-based networks involves three types of machines:
+**sentries**, **validators**, and **signers**.
+All of them play an essential role in the validation process.
+These machines can be located in different cites,
+and even in different countries.
+
+### Complex example: Ethereum anno 2024
+
+Since Ethereum moved to Proof of Stake in September 2022,
+operating a basic Ethereum validator requires multiple pieces of node software:
+
+ * An _execution client_, such as Geth.
+ * A _validator client_ for the consensus layer.
+ * A _beacon node_ for the consensus layer.
+ * _PBS middleware_, such as MEV-Boost.
```

Comment:
if we add "optional" software to the list perhaps it's worth adding remote signer such as web3signer? in our cases it's mandatory for all validator clients we run

Reply:
Ah yes, yet another one! And then behind web3signer is a database, which can of course be a distributed database like CockroachDB … although that is maybe one turtle too deep to mention here.

Reply:
yeah i'd keep database out of this :D but web3signer is worth adding